lib/std/os/windows.zig

```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
     return @bitCast(u64, result);
 }
 
+pub fn QueryObjectName(
+    handle: HANDLE,
+    out_buffer: []u16,
+) ![]u16 {
+    var full_buffer: [@sizeOf(OBJECT_NAME_INFORMATION) + PATH_MAX_WIDE * 2]u8 align(@alignOf(OBJECT_NAME_INFORMATION)) = undefined;
+    var info = @ptrCast(*OBJECT_NAME_INFORMATION, &full_buffer);
+    //buffer size is specified in bytes
+    const full_buffer_length = @intCast(ULONG, @sizeOf(OBJECT_NAME_INFORMATION) + std.math.min(PATH_MAX_WIDE, (out_buffer.len + 1) * 2));
+    //last argument would return the length required for full_buffer, not exposed here
+    const rc = ntdll.NtQueryObject(handle, .ObjectNameInformation, full_buffer[0..], full_buffer_length, null);
+    return switch (rc) {
+        .SUCCESS => if (@ptrCast(?[*]WCHAR, info.Name.Buffer)) |buffer| blk: {
```

Comment:
[The documentation for ObQueryNameString](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-obquerynamestring) , which I based this on since it seems to be used under the hood, mentions a `NULL` value under "Remarks" (`If the given object is unnamed [...]`).
`info.MaxLength` is also set to `0` in that case, so I'm happy to change the check to that, but returning an empty string would require checking for that in the consumer code (`GetFinalPathNameByHandle`).
The object's name isn't the empty string in this case, it has no name - I think either returning an optional or error still makes more sense.

Reply:
I'm still on the fence with the idea of having `error.Unexpected`. If something like this happens, shouldn't we simply trigger a nonrecoverable runtime trap instead? Or do we intend for user-space code to be recoverable in those circumstances? If that's the case, anyone's got any ideas for a scenario when that would be the case?

Reply:
"nonrecoverable runtime trap" (so basically `@panic`?) sounds very cut-throat to me.
If the file was supplied by a user, say interactively via `stdin`, maybe you just want to notify the user "this file didn't work, try another / moving it to a different path" (or something more specific if you figure out the underlying problem).
To me this seems like a fitting scenario for an error code: We failed to provide the requested service, but if/how to recover is up to the caller. That's why I made it return `error.Unnamed` before.
If we don't expect this API to be used with unnamed objects, then this scenario is "unexpected". `error.Unexpected` is not inherently reserved for kernel failures, so I think it fits.

But these are just my thoughts, whatever you think is best I'll implement.